### ISSUE TITLE
Mybinder bug in building master branch vs latest commit on the master branch

### ISSUE THREAD

**Issue by fzeiser** [jupyterhub/binder#173]
I have some odd behavior with the building of containers on mybinder.

I pushed some changes to my repo, after which mybinder was not able to build the docker image any longer. I fixed the issue and checked on my computer and dockerhub that the build is successful now. On the latter two it is, but mybinder still fails to build it.

The weird thing is: when I run this link: 
https://mybinder.org/v2/gh/oslocyclotronlab/ompy/master?filepath=ompy%2Fnotebooks%2Fgetting_started.ipynb
Binder fails to build the repo, with some error message related to the error before the latest commit. However, when I run mybinder and specify the lastest commit on master, the repo will be build.
https://mybinder.org/v2/gh/oslocyclotronlab/ompy/e7dd451c60b8d0e98a18c087cf8262cec5a99f0e.

I also tried just https://mybinder.org/v2/gh/oslocyclotronlab/ompy/master, but (as expected from the failure above) that doesn't work either

**Comment by fzeiser** [comment]
It seems like I get an unstable behavior with the https://mybinder.org/v2/gh/oslocyclotronlab/ompy/e7dd451c60b8d0e98a18c087cf8262cec5a99f0e link, now, too, eventhough I'm sure it build just some minutes ago. Anyhow, it was build on dockerhubs, so that's weird:
https://hub.docker.com/r/oslocyclotronlab/ompy/builds

**Comment by fzeiser** [comment]
The last part of the log file is also quite strange:
![image](https://user-images.githubusercontent.com/15156776/66572252-ef969b00-eb70-11e9-97f7-7011f16421bf.png)

This is ok for the error I've had before the latest commit (e7dd451c60b8d0e98a18c087cf8262cec5a99f0e), however, in this commit I changed some behavior in the `setup.py` file and added the last two lines in here:
``` py
    raise ValueError("Env var ompy_OpenMP must be either True or False "
                     "(or not set); use eg. 'export ompy_OpenMP=False'"
                     f"Now it is: {openmp}")
```
https://github.com/oslocyclotronlab/ompy/blob/e7dd451c60b8d0e98a18c087cf8262cec5a99f0e/setup.py#L125-L127

So, if the latest commit was build, I should receive the `"Now it is: {openmp}" in the `ValueError`, too, which I don't. Does mybinder lack behin in which commit it builds?

**Comment by minrk** [comment]
In attempting to test this PR, I think I found the answer. repo2docker does not have a stale checkout of the repo, but the *docker build cache* did. This is due to using `git clone` to fetch the repo contents in the Dockerfile instead of `COPY`. It is fixed in https://github.com/oslocyclotronlab/ompy/pull/58

**Comment by fzeiser** [comment]
I realized that COPY works for dockerhub just as it does for mybinder. Eventhough I don't quite understand form where docker gets the information about which repo to pull then.

I got a new problem now, with docker complaining about a file that could not be written in the repo I just cloned (or copied). This was actually just because of write permissions/ownership in the COPY command. I fixed it by specifying the correct user id and group:

COPY --chown=1000:100 . ompy

further explained in https://docs.docker.com/engine/reference/builder/#add.
I could probably also have used the suggestion from https://mybinder.readthedocs.io/en/latest/tutorials/dockerfile.html:

RUN chown -R ${NB_UID} ${HOME}

The final adjustment I needed to make in comparison to the recursive clone, was to add a git submodule update --init --recursive (as the example notebooks accesses files in a submodule).

Thanks again for the help!

**Comment by fzeiser** [comment]
Is this important enough to mention somewhere in the docs or shall we just close the issue and hope that people with a similar problem find this issue?

**Comment by betatim** [comment]
Our general policy on repos with `Dockerfiles` is that "if you use one we assume you are an expert who knows what they are doing". A bit like when you leave the marked trail in the mountains, you are probably doing it on purpose and have the required skills to do so. If not you should stick to the marked paths.

Not sure about extending the docs as step 4 [here](https://mybinder.readthedocs.io/en/latest/tutorials/dockerfile.html#preparing-your-dockerfile) already mentions `COPY`. Maybe a call out saying "a common mistake is to not use `COPY` but `RUN git clone ...` which leads to problems with caching". The problem with docs on `Dockerfile`s is that you can do what ever you want so it is really hard to write instructions/warnings because people use them to do custom stuff. Which is by definition ... custom to what they are doing :-/

**Comment by fzeiser** [comment]
I totally understand you point. I was thinking just in the direction of what you wrote 
> Maybe a call out saying "a common mistake is to not use COPY but RUN git clone ... which leads to problems with caching". 
as it is not a problem with my dockerfile (it worked perfectly fine on dockerhub), but led to an issue on mybinder.

**Comment by betatim** [comment]
Exactly. It is a valid `Dockerfile` but in combination with how BinderHub works it is something you can't/shouldn't do ... that is unless you want that behaviour for some reason. I think the best way forward is creating a PR and then discussing based on that.
